src/monogame/vulkan/MGG_Vulkan.cpp

```diff
@@ -0,0 +1,973 @@
+};
+
+struct MGG_InputLayout
+{
+	MGG_Shader* vertexShader;
+};
+
+struct MGG_Shader
+{
+	MGShaderStage stage;
+	mgbyte* bytecode;
+	mgint bytecodeSize;
+};
+
+struct MGG_BlendState
+{
+	MGG_BlendState_Info info[4];
+};
+
+struct MGG_DepthStencilState
+{
+	MGG_DepthStencilState_Info info;
+};
+
+struct MGG_RasterizerState
+{
+	MGG_RasterizerState_Info info;
+};
+
+struct MGG_SamplerState
+{
+	MGG_SamplerState_Info info;
+};
+
+
+struct MGG_System
+{
+	VkInstance instance;
+
+	std::vector<MGG_GraphicsAdapter*> adapters;
+};
+
+static MGG_System* mgg_system = nullptr;
+
+
+void MGG_EffectResource_GetBytecode(const char* name, mgbyte** bytecode, mgint* size)
+{
+	*bytecode = nullptr;
+	*size = 0;
+
+	// Get the handle of this DLL.
+	HMODULE module;
+	GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, (LPCSTR)&MGG_EffectResource_GetBytecode, &module);
+
+	LPCSTR id = nullptr;
+
+	if (strcmp(name, "AlphaTestEffect") == 0)
+		id = MAKEINTRESOURCEA(C_AlphaTestEffect);
+	else if (strcmp(name, "BasicEffect") == 0)
+		id = MAKEINTRESOURCEA(C_BasicEffect);
+	else if (strcmp(name, "DualTextureEffect") == 0)
+		id = MAKEINTRESOURCEA(C_DualTextureEffect);
+	else if (strcmp(name, "EnvironmentMapEffect") == 0)
+		id = MAKEINTRESOURCEA(C_EnvironmentMapEffect);
+	else if (strcmp(name, "SkinnedEffect") == 0)
+		id = MAKEINTRESOURCEA(C_SkinnedEffect);
+	else if (strcmp(name, "SpriteEffect") == 0)
+		id = MAKEINTRESOURCEA(C_SpriteEffect);
+
+	auto handle = ::FindResourceA(module, id, "BIN");
+	if (handle == nullptr)
+		return;
+
+	*size = SizeofResource(module, handle);
+	if (*size == 0)
+		return;
+
+	HGLOBAL global = LoadResource(module, handle);
+	if (global == nullptr)
+		return;
+
+	*bytecode = (mgbyte*)LockResource(global);
+}
+
+void System_GetAdapters(MGG_System* system)
+{
+	uint32_t count = 0;
+	VkResult res = vkEnumeratePhysicalDevices(system->instance, &count, NULL);
+	if (res != VK_SUCCESS)
+		return;
+
+	VkPhysicalDevice* gpus = (VkPhysicalDevice*)calloc(count, sizeof(*gpus));
+
+	res = vkEnumeratePhysicalDevices(system->instance, &count, gpus);
+	if (res == VK_SUCCESS)
+	{
+		for (uint32_t i = 0; i < count; ++i)
+		{
+			auto adapter = new MGG_GraphicsAdapter();
+			adapter->device = gpus[i];
+
+			vkGetPhysicalDeviceProperties(adapter->device, &adapter->properties);
+			vkGetPhysicalDeviceFeatures(adapter->device, &adapter->features);
+			vkGetPhysicalDeviceMemoryProperties(adapter->device, &adapter->memory);
+
+			system->adapters.push_back(adapter);
+		}
+	}
+
+	free((void*)gpus);
+}
+
+static MGG_System* System_Get()
+{
+	if (mgg_system != nullptr)
+		return mgg_system;
+
+	auto err = volkInitialize();
+	if (err != VK_SUCCESS)
+	{
+		printf("Failed to initialize volk!");
+		return nullptr;
+	}
+
+	VkApplicationInfo app_info = { VK_STRUCTURE_TYPE_APPLICATION_INFO };
+	app_info.pNext = nullptr;
+	app_info.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
+	app_info.pApplicationName = "None";
```

Comment:
Ideally we pass down an application name...  i assume this is used by drivers and utilities for optimizing/tweaking settings for specific games.

Reply:
Yes, vulkan has fields to pass both game name and engine name so the driver can optimise it if it so chooses.